sklearn/tests/test_metaestimators_metadata_routing.py

```diff
@@ -875,3 +877,28 @@ def test_metadata_is_routed_correctly_to_splitter(metaestimator):
             check_recorded_metadata(
                 obj=_splitter, method="split", parent=method_name, **method_kwargs
             )
+
+
+@pytest.mark.parametrize("metaestimator", METAESTIMATORS, ids=METAESTIMATOR_IDS)
+def test_metadata_routed_to_group_splitter(metaestimator):
+    """Test that groups are routed correctly if group splitter of CV estimator is used
+    within cross_validate."""
+
+    if "cv_routing_methods" not in metaestimator:
+        # This test is only for metaestimators accepting a CV splitter
+        return
+
+    cls = metaestimator["metaestimator"]
+    X_ = metaestimator["X"]
+    y_ = metaestimator["y"]
+
+    try:
+        cross_validate(
+            cls(cv=GroupKFold(n_splits=2)),
+            X_,
+            y_,
+            params={"groups": groups},
+            cv=GroupKFold(n_splits=2),
+        )
+    except TypeError:
```

Comment:
when does this happen? looks suspicious

Reply:
I agree that this seems confusing. Either we should try to avoid the exception or maybe add a comment stating why it occurs if it is not really related to the functionality being tested.

Reply:
This is why I called my test clumsy in the description. The TypeErrors are completely unrelated:

```
FAILED sklearn/tests/test_metaestimators_metadata_routing.py::test_metadata_routed_to_group_splitter[GridSearchCV] - TypeError: GridSearchCV.__init__() missing 2 required positional arguments: 'estimator' and 'param_grid'
FAILED sklearn/tests/test_metaestimators_metadata_routing.py::test_metadata_routed_to_group_splitter[RandomizedSearchCV] - TypeError: RandomizedSearchCV.__init__() missing 2 required positional arguments: 'estimator' and 'param_distributions'
FAILED sklearn/tests/test_metaestimators_metadata_routing.py::test_metadata_routed_to_group_splitter[HalvingGridSearchCV] - TypeError: HalvingGridSearchCV.__init__() missing 2 required positional arguments: 'estimator' and 'param_grid'
FAILED sklearn/tests/test_metaestimators_metadata_routing.py::test_metadata_routed_to_group_splitter[HalvingRandomSearchCV] - TypeError: HalvingRandomSearchCV.__init__() missing 2 required positional arguments: 'estimator' and 'param_distributions'
FAILED sklearn/tests/test_metaestimators_metadata_routing.py::test_metadata_routed_to_group_splitter[SequentialFeatureSelector] - TypeError: SequentialFeatureSelector.__init__() missing 1 required positional argument: 'estimator'
FAILED sklearn/tests/test_metaestimators_metadata_routing.py::test_metadata_routed_to_group_splitter[RFECV] - TypeError: RFECV.__init__() missing 1 required positional argument: 'estimator'
```

Reply:
Maybe we can remove these from metaestimators from this test? Most of them seem to be related to *SearchCV and the last two require an "estimator" in the initialization.

Reply:
I would rather fix the test infrastructure to test the `*SearchCV` estimators as those are typical cases where we want to make sure that CV related metadata are routed correctly.

Maybe we could reuse `_contruct_instance` here. Instead of:

```python
est_class(cv=GroupKFold(n_splits=2))
```

for something like:

```python
est = next(_construct_instances(metaestimator["metaestimator"]))
est.set_params(cv=GroupKFold(n_splits=2))
```

Note that I think we should rather avoid naming local variables `cls` because `cls` is a specific keyword used in the context of class methods by convention. If you really want to use a local variable, I would rather name it something more specific as `est_class` or `metaestimator_class`.

Also note that we need `next` in the snippet above, because `_construct_instance` is a generator that can return several instances (by iterating over it) since #29824 was merged.